### PR TITLE
awscrt.auth integration for botocore v2

### PR DESCRIPTION
**Pull request by graebm**
- Implement new SigV4 signers based on `awscrt.auth` (The old SigV4 signers are still the default)
- Replace existing SigV4 test suite with updated one.

Notes:
- Because awscrt requires Python 3.5+, the Travis tests fail in the Python 2.7 and 3.4 runs.
- Future work will be:
  - make awscrt an optional dependency
  - use CRT signers if awscrt is present

### CODE REVIEW

File: botocore/auth.py [hunk 1]
```diff
@@ -615,6 +628,285 @@ def add_auth(self, request):
...
+            # need to create new url with the new query string.
+            # <part>   - <index>
+            # scheme   - 0
+            # netloc   - 1
+            # path     - 2
+            # query    - 3  <-- we're replacing this.
+            # fragment - 4
+            aws_request.url = urlunsplit(
+                (p[0], p[1], p[2], new_query_string, p[4]))
+
```
Comment (kyleknap) [inline botocore/auth.py:698]: Instead of putting an if statement specific to query signing in the main sigv4 signer, it may make sense to just override the `_apply_signing_changes()` method in the `CrtSigV4QueryAuth`. Otherwise, its a little hard to follow why this class has query signing logic and it would be nice to isolate all of the query specific logic to one class.

**Review by kyleknap — COMMENTED** [review]
Looks good! Did not really have any big feedback. Just some suggestions and questions.

File: setup.cfg [hunk 2]
```diff
@@ -3,6 +3,7 @@ universal = 1
 
 [metadata]
 requires-dist =
+    awscrt>=0.9,<0.10
```
[the remark this answers is not in the document]
Comment (graebm) [reply setup.cfg:6]: I'll pin this to a specific patch.

Unfortunately, while the major version is 0 we're still sometimes making breaking API changes (auth changed several times over the course of working on this).

File: tests/unit/auth/test_sigv4.py [hunk 3]
```diff
@@ -140,25 +141,39 @@ def create_request_from_raw_request(raw_request):
...
-def _test_signature_version_4(test_case):
+def _test_signature_version_4(test_case, signer):
     test_case = _SignatureTestCase(test_case)
     request = create_request_from_raw_request(test_case.raw_request)
 
-    auth = botocore.auth.SigV4Auth(test_case.credentials, 'host', 'us-east-1')
+    if signer == 'crt':
+        # Use CRT logging to diagnose interim steps (canonical request, etc)
+        # import awscrt.io
+        # awscrt.io.init_logging(awscrt.io.LogLevel.Trace, 'stdout')
```
[the remark this answers is not in the document]
Comment (graebm) [reply tests/unit/auth/test_sigv4.py:151]: We'd have to do a bit of work pipe this through the standard library logger. I've always wanted to do this, but haven't had the time to experiment with doing it the right way, so I'm glad you're formally requesting it.

File: botocore/auth.py [hunk 4]
```diff
@@ -615,6 +628,285 @@ def add_auth(self, request):
...
+        # bit of metadata through the request context.
+        return request.context.get('payload_signing_enabled', True)
+
+    def _should_add_content_sha256_header(self, explicit_payload):
+        # only add X-Amz-Content-SHA256 header if payload is explicitly set
+        return explicit_payload is not None
+
+    def add_auth(self, request):
+        if self.credentials is None:
+            raise NoCredentialsError
```
[the remark this answers is not in the document]
Comment (graebm) [reply botocore/auth.py:731]: good catch.
I totally copy/pasted this

File: tests/unit/auth/test_sigv4.py [hunk 5]
```diff
@@ -42,27 +43,29 @@
...
     'get-vanilla-query-order-key-case',
-    # Multiple query params of the same key not supported by
-    # the SDKs.
+    'get-vanilla-query-order-key',
     'get-vanilla-query-order-value',
+    # Test seems wrong: multiline header value should not get commas inserted.
+    'get-header-value-multiline',
+    # Test seems wrong: String-to-sign's hash of the canonical request is off.
+    'post-x-www-form-urlencoded',
+    'post-x-www-form-urlencoded-parameters',
```
[the remark this answers is not in the document]
Comment (graebm) [reply tests/unit/auth/test_sigv4.py:68]: I copied over the fixed `post-x-www-form-urlencoded` and `post-x-www-form-urlencoded-parameters` from the Ruby SDK, and removed these from the ignore list.

The `get-vanilla-query-order-XYZ` tests should stay in the ignore list. These tests have repeated keys in the query-params which is something Botocore just doesn't support.

File: tests/unit/auth/test_sigv4.py [hunk 6]
```diff
@@ -42,27 +43,29 @@
...
-    # the SDKs.
+    # Bad request-line syntax, python's HTTP parser chokes on this.
+    'normalize-path/get-space',
+    # Multiple query params of the same key not supported by the SDKs.
     'get-vanilla-query-order-key-case',
-    # Multiple query params of the same key not supported by
-    # the SDKs.
+    'get-vanilla-query-order-key',
     'get-vanilla-query-order-value',
+    # Test seems wrong: multiline header value should not get commas inserted.
```
[the remark this answers is not in the document]
Comment (graebm) [reply tests/unit/auth/test_sigv4.py:64]: Botocore is doing the right thing. This test is wrong, and the CRT is doing the wrong thing to pass this test. A newline in a header value should be canonicalized with a space, not a comma.

I hacked up an integration test to confirm this. I hit S3 with a request that had a multiline header in it. If I canonicalized it with commas, S3 rejected it and the response showed a canonical request with spaces. When I canonicalized the request with commas S3 accepted the request.

I notice Ruby did *not* alter their multiline test, so they might be making the same mistake as the CRT and inserting commas.

[the remark this answers is not in the document]
Comment (graebm) [reply tests/unit/auth/test_sigv4.py:64]: Looked at Ruby more... the code that parses the test files pretends the multiline header is a repeated header. So that's how they pass the test. Not sure what happens if they encounter a multiline header in the wild.

[the remark this answers is not in the document]
Comment (graebm) [reply tests/unit/auth/test_sigv4.py]: Fixed test data. Fixed CRT doing it wrong. Using latest CRT where we're not doing it wrong anymore.

**Review by kyleknap — APPROVED** [review]
Looks good! :ship:
